package com.antik.DexPatcher.MethodT;

import com.android.tools.smali.dexlib2.Opcode;
import com.android.tools.smali.dexlib2.iface.ExceptionHandler;
import com.android.tools.smali.dexlib2.iface.Method;
import com.android.tools.smali.dexlib2.iface.MethodImplementation;
import com.android.tools.smali.dexlib2.iface.TryBlock;
import com.android.tools.smali.dexlib2.iface.instruction.Instruction;
import com.android.tools.smali.dexlib2.iface.instruction.ReferenceInstruction;
import com.android.tools.smali.dexlib2.iface.reference.MethodReference;
import com.android.tools.smali.dexlib2.immutable.ImmutableExceptionHandler;
import com.android.tools.smali.dexlib2.immutable.ImmutableMethod;
import com.android.tools.smali.dexlib2.immutable.ImmutableMethodImplementation;
import com.android.tools.smali.dexlib2.immutable.ImmutableTryBlock;
import com.android.tools.smali.dexlib2.immutable.instruction.ImmutableInstruction35c;
import com.android.tools.smali.dexlib2.immutable.reference.ImmutableMethodReference;

import java.util.ArrayList;
import java.util.List;

public class patchLaunchMethod {
    public static Method patchLaunchMethod(Method m) {
        MethodImplementation im = m.getImplementation();
        if (im == null) return m;
        List<Instruction> o_ins = new ArrayList<>();
        for (Instruction i : im.getInstructions()) {
        o_ins.add(i);
        }
        List<Instruction> n_ins = new ArrayList<>();
        int p_off = -1;
        int c_off = 0;

        for (int i = 0; i < o_ins.size(); i++) {
            Instruction inst = o_ins.get(i);
            n_ins.add(inst);
            int u = inst.getCodeUnits();
            if (inst instanceof ReferenceInstruction r_i) {
                if (r_i.getReference() instanceof MethodReference m_r) {
                    if ("Lcom/pairip/VMRunner;".equals(m_r.getDefiningClass()) && "invoke".equals(m_r.getName())) {

                        boolean a_p = false;
                        if (i + 1 < o_ins.size()) {
                            Instruction nx = o_ins.get(i + 1);
                            if (nx instanceof ReferenceInstruction n_r) {
                                if (n_r.getReference() instanceof MethodReference n_m) {
                                    if ("pairip".equals(n_m.getName())) {
                                        a_p = true;
                                    }
                                }
                            }
                        }
                        if (!a_p) {
                            p_off = c_off + u;
                            n_ins.add(new ImmutableInstruction35c(Opcode.INVOKE_STATIC, 0, 0, 0, 0, 0, 0, new ImmutableMethodReference("Lcom/pairip/StartupLauncher;", "pairip", null, "V")));
                        }
                    }
                }
            }
            c_off += u;
        }

        if (p_off == -1) return m;

        List<? extends TryBlock<? extends ExceptionHandler>> o_tbs = im.getTryBlocks();
        List<ImmutableTryBlock> n_tbs = new ArrayList<>();

        for (TryBlock<? extends ExceptionHandler> t : o_tbs) {
            int s = t.getStartCodeAddress();
            int c = t.getCodeUnitCount();
            int e = s + c;
            List<ImmutableExceptionHandler> h = new ArrayList<>();
            for (ExceptionHandler ha : t.getExceptionHandlers()) {
                int h_a = ha.getHandlerCodeAddress();
                if (h_a >= p_off) {
                    h.add(new ImmutableExceptionHandler(ha.getExceptionType(), h_a + 3));
                } else {
                    h.add(ImmutableExceptionHandler.of(ha));
                }
            }
            if (s < p_off && e >= p_off) {
                n_tbs.add(new ImmutableTryBlock(s, c + 3, h));
            } else if (s >= p_off) {
                n_tbs.add(new ImmutableTryBlock(s + 3, c, h));
            } else {
                n_tbs.add(new ImmutableTryBlock(s, c, h));
            }
        }

        return new ImmutableMethod(m.getDefiningClass(), m.getName(), m.getParameters(), m.getReturnType(), m.getAccessFlags(), m.getAnnotations(), m.getHiddenApiRestrictions(), new ImmutableMethodImplementation(im.getRegisterCount(), n_ins, n_tbs, im.getDebugItems()));
    }
}
